File: frontline-app/lib/services/triage_service.dart
import '../models/fhir_models.dart';

class TriageRule {
  final String id;
  final String version;
  final String appliesTo;
  final String riskBand;
  final String flag;
  final String rationale;
  final String source;
  final String effectiveDate;
  final String recommendedAction;
  final bool Function(Map<String, double> vars) evaluate;

  TriageRule({
    required this.id,
    required this.version,
    required this.appliesTo,
    required this.riskBand,
    required this.flag,
    required this.rationale,
    required this.source,
    required this.effectiveDate,
    required this.recommendedAction,
    required this.evaluate,
  });
}

class TriageResult {
  final String riskBand;
  final List<String> flags;
  final List<Map<String, String>> triggeredRules;
  final String recommendedAction;
  final String protocolVersion;

  TriageResult({
    required this.riskBand,
    required this.flags,
    required this.triggeredRules,
    required this.recommendedAction,
    required this.protocolVersion,
  });
}

class TriageService {
  final String _currentProtocolVersion = 'v1.1.0';
  
  final List<TriageRule> _rules = [
    TriageRule(
      id: 'rule-general-spo2',
      version: '1.0.0',
      appliesTo: 'General',
      riskBand: 'EMERGENCY',
      flag: 'Severe Hypoxia (SpO2 < 90%)',
      rationale: 'SpO2 below 90% indicates severe respiratory distress requiring immediate oxygen therapy.',
      source: 'WHO Clinical Management of COVID-19',
      effectiveDate: '2023-01-01',
      recommendedAction: 'IMMEDIATE_ESCALATION',
      evaluate: (vars) => vars.containsKey('spo2') && vars['spo2']! < 90,
    ),
    TriageRule(
      id: 'rule-anc-bp',
      version: '1.1.0',
      appliesTo: 'ANC',
      riskBand: 'EMERGENCY',
      flag: 'Severe Hypertension / Suspected Pre-eclampsia',
      rationale: 'Systolic >= 160 or Diastolic >= 110 in pregnancy is a hypertensive crisis.',
      source: 'ACOG Practice Bulletin No. 222',
      effectiveDate: '2023-06-01',
      recommendedAction: 'IMMEDIATE_ESCALATION',
      evaluate: (vars) => (vars.containsKey('bp.systolic') && vars['bp.systolic']! >= 160) || 
                          (vars.containsKey('bp.diastolic') && vars['bp.diastolic']! >= 110),
    ),
    TriageRule(
      id: 'rule-anc-bp-high',
      version: '1.0.1',
      appliesTo: 'ANC',
      riskBand: 'HIGH_RISK',
      flag: 'Gestational Hypertension',
      rationale: 'Systolic >= 140 or Diastolic >= 90 indicates gestational hypertension requiring monitoring.',
      source: 'ACOG Practice Bulletin No. 222',
      effectiveDate: '2023-01-01',
      recommendedAction: 'PHC_CONSULT_WITHIN_24H',
      evaluate: (vars) => (vars.containsKey('bp.systolic') && vars['bp.systolic']! >= 140) || 
                          (vars.containsKey('bp.diastolic') && vars['bp.diastolic']! >= 90),
    ),
  ];

  TriageResult evaluate(String encounterType, List<Observation> observations) {
    Map<String, double> variables = {};
    for (var obs in observations) {
      if (obs.code != null && obs.value != null) {
        variables[obs.code!] = (obs.value as num).toDouble();
      }
    }

    String finalRiskBand = 'ROUTINE';
    List<String> flags = [];
    List<Map<String, String>> triggeredRules = [];
    String recommendedAction = 'ROUTINE_CARE';

    for (var rule in _rules) {
      if (rule.appliesTo == encounterType || rule.appliesTo == 'General') {
        if (rule.evaluate(variables)) {
          flags.add(rule.flag);
          triggeredRules.add({
            'ruleId': rule.id,
            'version': rule.version,
            'rationale': rule.rationale,
            'source': rule.source,
          });
          
          if (rule.riskBand == 'EMERGENCY') {
            finalRiskBand = 'EMERGENCY';
            recommendedAction = rule.recommendedAction;
          } else if (rule.riskBand == 'HIGH_RISK' && finalRiskBand != 'EMERGENCY') {
            finalRiskBand = 'HIGH_RISK';
            recommendedAction = rule.recommendedAction;
          }
        }
      }
    }

    return TriageResult(
      riskBand: finalRiskBand,
      flags: flags,
      triggeredRules: triggeredRules,
      recommendedAction: recommendedAction,
      protocolVersion: _currentProtocolVersion,
    );
  }
}

File: core-gateway/api/src/triage/triage.service.ts
import { Injectable, Logger, NotFoundException } from '@nestjs/common';
import * as jsonLogic from 'json-logic-js';
import { FhirService } from '../fhir/fhir.service.js';
import { AuditService } from '../audit/audit.service.js';
import { v4 as uuidv4 } from 'uuid';

@Injectable()
export class TriageService {
  private readonly logger = new Logger(TriageService.name);

  // Versioned Protocol Rules
  private readonly RULE_VERSION = 'v1.1.0';
  private rules = [
    {
      "id": "rule-general-spo2",
      "appliesTo": "General",
      "condition": { "<": [{ "var": "spo2" }, 90] },
      "outcome": {
        "riskBand": "EMERGENCY",
        "flag": "Severe Hypoxia (SpO2 < 90%)",
        "recommendedAction": "IMMEDIATE_ESCALATION"
      }
    },
    {
      "id": "rule-anc-bp",
      "appliesTo": "ANC",
      "condition": {
        "or": [
          { ">=": [{ "var": "bp.systolic" }, 160] },
          { ">=": [{ "var": "bp.diastolic" }, 110] }
        ]
      },
      "outcome": {
        "riskBand": "EMERGENCY",
        "flag": "Severe Hypertension / Suspected Pre-eclampsia",
        "recommendedAction": "IMMEDIATE_ESCALATION"
      }
    },
    {
      "id": "rule-anc-bp-high",
      "appliesTo": "ANC",
      "condition": {
        "or": [
          { ">=": [{ "var": "bp.systolic" }, 140] },
          { ">=": [{ "var": "bp.diastolic" }, 90] }
        ]
      },
      "outcome": {
        "riskBand": "HIGH_RISK",
        "flag": "Gestational Hypertension",
        "recommendedAction": "PHC_CONSULT_WITHIN_24H"
      }
    }
  ];

  constructor(
    private readonly fhirService: FhirService,
    private readonly auditService: AuditService
  ) {}

  async evaluateEncounter(encounterId: string, user: any, correlationId: string) {
    this.logger.log(`Evaluating triage for encounter \${encounterId}`);
    
    const encounter = await this.fhirService.getResource('Encounter', encounterId);
    if (!encounter) throw new NotFoundException('Encounter not found');
    
    const patientRef = encounter.subject?.reference;
    if (!patientRef) throw new NotFoundException('Encounter has no subject');
    const patientId = patientRef.split('/')[1];

    // Fetch Observations linked to this encounter
    const obsBundle = await this.fhirService.getResource('Observation', `?encounter=Encounter/\${encounterId}`);
    const observations = obsBundle?.entry?.map(e => e.resource) || [];

    // Extract values
    const variables: Record<string, number> = {};
    for (const obs of observations) {
      if (obs.resourceType === 'Observation' && obs.code?.coding?.[0]?.code) {
        const code = obs.code.coding[0].code;
        if (obs.valueQuantity?.value !== undefined) {
          variables[code] = obs.valueQuantity.value;
        }
      }
    }

    let finalRiskBand = 'ROUTINE';
    const flags = [];
    let recommendedAction = 'ROUTINE_CARE';
    const ruleTrace = [];

    // Evaluate
    // For demo, if class is ANC, we use ANC rules. Otherwise General.
    const encounterType = encounter.class?.code === 'ANC' ? 'ANC' : 'General';

    for (const rule of this.rules) {
      if (rule.appliesTo === encounterType || rule.appliesTo === 'General') {
        const isMatch = jsonLogic.apply((rule as any).condition, variables);
        if (isMatch) {
          ruleTrace.push(rule.id);
          flags.push(rule.outcome.flag);
          
          if (rule.outcome.riskBand === 'EMERGENCY') {
            finalRiskBand = 'EMERGENCY';
            recommendedAction = rule.outcome.recommendedAction;
          } else if (rule.outcome.riskBand === 'HIGH_RISK' && finalRiskBand !== 'EMERGENCY') {
            finalRiskBand = 'HIGH_RISK';
            recommendedAction = rule.outcome.recommendedAction;
          }
        }
      }
    }

    // Generate FHIR RiskAssessment
    const riskId = uuidv4();
    const riskAssessment = {
      resourceType: 'RiskAssessment',
      id: riskId,
      status: 'final',
      subject: { reference: `Patient/\${patientId}` },
      encounter: { reference: `Encounter/\${encounterId}` },
      method: { coding: [{ system: 'http://setu.in/protocols', code: this.RULE_VERSION }] },
      prediction: [{
        qualitativeRisk: { text: finalRiskBand },
        rationale: flags.join('; ') || 'No elevated risk indicators detected.'
      }],
      mitigation: recommendedAction
    };

    await this.fhirService.createOrUpdate('RiskAssessment', riskId, riskAssessment, undefined, 'CREATE');

    await this.auditService.logEvent({
      userId: user.userId,
      role: user.role,
      facilityId: user.facilityId,
      action: 'TRIAGE_COMPLETED',
      resourceType: 'RiskAssessment',
      resourceId: riskId,
      requestId: correlationId,
      result: 'SUCCESS',
    });

    if (finalRiskBand === 'EMERGENCY' || finalRiskBand === 'HIGH_RISK') {
      await this.auditService.logEvent({
        userId: user.userId,
        role: user.role,
        facilityId: user.facilityId,
        action: 'RISK_IDENTIFIED',
        resourceType: 'RiskAssessment',
        resourceId: riskId,
        requestId: correlationId,
        result: 'SUCCESS',
      });
    }

    return riskAssessment;
  }
}
